File: node/src/core/suppression-writer.ts
import fs from "fs";
import path from "path";
import { execSync } from "child_process";
import yaml from "js-yaml";
import { findPolicyFile } from "./policy-loader.js";

export interface SuppressionInput {
  /** File path or glob to suppress findings in. Required, non-empty. */
  paths: string[];
  /** Specific rule/pattern names to suppress. Omitted/empty = suppress all rules for those paths. */
  rules?: string[];
  /** Human-readable rationale, persisted alongside the rule. */
  reason?: string;
  /** Base directory for resolving the policy file. Defaults to process.cwd(). */
  cwd?: string;
}

export interface SuppressionResult {
  /** Absolute path of the policy file written. */
  file: string;
  /** What happened: a new file was created, a rule appended, or an existing rule's reason updated. */
  action: "created" | "appended" | "updated";
  /** The ignore rule as persisted. */
  entry: { paths: string[]; rules?: string[]; reason?: string };
  /** Total number of ignore rules in the file after the write. */
  suppressionCount: number;
}

function getGitRoot(cwd: string): string | null {
  try {
    return execSync("git rev-parse --show-toplevel", {
      cwd,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "ignore"],
    }).trim();
  } catch {
    return null;
  }
}

/**
 * Normalize a rule for dedup comparison: an ignore rule is "the same" if it
 * targets the same set of paths and the same set of rule names (order- and
 * duplicate-insensitive). Reason is intentionally excluded — re-suppressing
 * the same scope just updates the reason.
 */
function ruleKey(paths: string[], rules?: string[]): string {
  const norm = (xs?: string[]) =>
    Array.from(new Set((xs ?? []).map((x) => String(x)))).sort();
  return JSON.stringify({ paths: norm(paths), rules: norm(rules) });
}

/**
 * Persist a finding suppression into the project's `.rafter.yml` `ignore`
 * list. Resolves the policy file via the same precedence the loader uses;
 * if none exists, creates a canonical `.rafter.yml` at the git root (or cwd).
 *
 * Merge semantics: if an existing ignore rule targets the same paths + rules,
 * its reason is updated in place rather than appending a duplicate.
 */
export function writeSuppression(input: SuppressionInput): SuppressionResult {
  const paths = (input.paths ?? []).map((p) => String(p)).filter((p) => p.length > 0);
  if (paths.length === 0) {
    throw new Error('"paths" must be a non-empty array of file paths or globs.');
  }
  const rules = Array.isArray(input.rules)
    ? input.rules.map((r) => String(r)).filter((r) => r.length > 0)
    : undefined;
  const reason = typeof input.reason === "string" && input.reason.trim() ? input.reason.trim() : undefined;
  const baseDir = input.cwd || process.cwd();

  // Resolve target file: existing policy file wins; else canonical dotfile at git root / cwd.
  let target = findPolicyFile(baseDir);
  let action: SuppressionResult["action"];
  let raw: Record<string, any> = {};

  if (target && fs.existsSync(target)) {
    const content = fs.readFileSync(target, "utf-8");
    const parsed = yaml.load(content);
    raw = parsed && typeof parsed === "object" ? (parsed as Record<string, any>) : {};
    action = "appended";
  } else {
    const root = getGitRoot(baseDir) || baseDir;
    target = path.join(root, ".rafter.yml");
    action = "created";
  }

  const ignoreList: any[] = Array.isArray(raw.ignore) ? raw.ignore : [];

  const newEntry: { paths: string[]; rules?: string[]; reason?: string } = { paths };
  if (rules && rules.length > 0) newEntry.rules = rules;
  if (reason) newEntry.reason = reason;

  const key = ruleKey(paths, rules);
  const existing = ignoreList.find(
    (e) => e && typeof e === "object" && Array.isArray(e.paths) && ruleKey(e.paths, e.rules) === key,
  );

  if (existing) {
    // Same scope already suppressed — update the reason in place.
    if (reason) existing.reason = reason;
    else delete existing.reason;
    if (action !== "created") action = "updated";
  } else {
    ignoreList.push(newEntry);
  }

  raw.ignore = ignoreList;

  const dumped = yaml.dump(raw, { lineWidth: 100, noRefs: true, sortKeys: false });
  const dir = path.dirname(target);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(target, dumped, "utf-8");

  return {
    file: target,
    action,
    entry: existing ? { ...newEntry, ...(existing.reason ? { reason: existing.reason } : {}) } : newEntry,
    suppressionCount: ignoreList.length,
  };
}

File: python/rafter_cli/core/suppression_writer.py
"""Persist finding suppressions into the project's .rafter.yml ignore list.

Mirrors node/src/core/suppression-writer.ts — keep both in sync.
"""
from __future__ import annotations

import json
from pathlib import Path

from ..utils.git import get_git_root
from .policy_loader import find_policy_file


def _rule_key(paths: list[str], rules: list[str] | None) -> str:
    """Order- and duplicate-insensitive identity for an ignore rule.

    Two rules are "the same" if they target the same set of paths and the
    same set of rule names. Reason is excluded — re-suppressing the same
    scope just updates the reason.
    """
    def norm(xs: list[str] | None) -> list[str]:
        return sorted({str(x) for x in (xs or [])})

    return json.dumps({"paths": norm(paths), "rules": norm(rules)}, sort_keys=True)


def write_suppression(
    paths: list[str],
    rules: list[str] | None = None,
    reason: str | None = None,
    cwd: str | Path | None = None,
) -> dict:
    """Persist a finding suppression into the project's .rafter.yml ignore list.

    Resolves the policy file via the same precedence the loader uses; if none
    exists, creates a canonical ``.rafter.yml`` at the git root (or ``cwd``).

    Merge semantics: if an existing ignore rule targets the same paths + rules,
    its reason is updated in place rather than appending a duplicate.

    ``cwd`` overrides the base directory for resolution (defaults to the process cwd).

    Returns a dict: {file, action, entry, suppression_count}.
    """
    import yaml

    norm_paths = [str(p) for p in (paths or []) if str(p)]
    if not norm_paths:
        raise ValueError('"paths" must be a non-empty list of file paths or globs.')
    norm_rules = [str(r) for r in rules if str(r)] if isinstance(rules, list) else None
    norm_reason = reason.strip() if isinstance(reason, str) and reason.strip() else None
    base_dir = str(cwd) if cwd else str(Path.cwd())

    target = find_policy_file(base_dir)
    raw: dict = {}

    if target and Path(target).exists():
        parsed = yaml.safe_load(Path(target).read_text())
        raw = parsed if isinstance(parsed, dict) else {}
        action = "appended"
    else:
        root = get_git_root() or base_dir
        target = Path(root) / ".rafter.yml"
        action = "created"

    target = Path(target)
    ignore_list = raw.get("ignore") if isinstance(raw.get("ignore"), list) else []

    new_entry: dict = {"paths": norm_paths}
    if norm_rules:
        new_entry["rules"] = norm_rules
    if norm_reason:
        new_entry["reason"] = norm_reason

    key = _rule_key(norm_paths, norm_rules)
    existing = next(
        (
            e
            for e in ignore_list
            if isinstance(e, dict)
            and isinstance(e.get("paths"), list)
            and _rule_key(e["paths"], e.get("rules")) == key
        ),
        None,
    )

    if existing is not None:
        if norm_reason:
            existing["reason"] = norm_reason
        else:
            existing.pop("reason", None)
        if action != "created":
            action = "updated"
        entry = dict(new_entry)
        if existing.get("reason"):
            entry["reason"] = existing["reason"]
    else:
        ignore_list.append(new_entry)
        entry = new_entry

    raw["ignore"] = ignore_list

    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(yaml.safe_dump(raw, sort_keys=False, default_flow_style=False))

    return {
        "file": str(target),
        "action": action,
        "entry": entry,
        "suppression_count": len(ignore_list),
    }
